tests/generative/variable_strategies.py

```diff
@@ -7,317 +8,212 @@
+#     * `@st.composite` allows us to define a strategy by composing other strategies with
+#       arbitrary Python code; it adds a `draw` argument which is part of the machinery that
+#       enables this composition but which doesn't form part of the signature of the resulting
+#       strategy function.
+
+
+def variable(patient_tables, event_tables, schema, value_strategies):
+    # Every inner-function here returns a Hypothesis strategy for creating the thing it is named
+    # for, not the thing itself.
+    #
+    # Several of these strategy functions ignore one or more of their arguments in order to make
+    # them uniform with other functions that return the same sort of strategy. Such ignored
+    # arguments are named with a leading underscore.
+
+    # Series strategies
+    #
+    # Whenever a series is needed, we call series() passing the type of the series and frame that
+    # it should be built on (these are either constrained by the context in which the series is to
+    # be used or chosen arbitrarily by the caller).
+    #
+    # This strategy then chooses an arbitrary concrete series that respects the constraints imposed
+    # by the passed type and frame.
+    @st.composite
+    def series(draw, type_, frame):
+        class DomainConstraint:
+            PATIENT = (True,)
+            NON_PATIENT = (False,)
+            ANY = (True, False)
+
+        # Order matters: "simpler" first (see header comment)
+        series_constraints = {
+            value: ({int, float, bool, datetime.date}, DomainConstraint.PATIENT),
+            select_column: ({int, float, bool, datetime.date}, DomainConstraint.ANY),
+            eq: ({bool}, DomainConstraint.ANY),
+            add: ({int, float}, DomainConstraint.ANY),
+            count: ({int}, DomainConstraint.PATIENT),
+        }
+        series_types = series_constraints.keys()
+
+        def constraints_match(series_type):
+            type_constraint, domain_constraint = series_constraints[series_type]
+            return (
+                type_ in type_constraint
+                and is_one_row_per_patient_frame(frame) in domain_constraint
+            )
+
+        possible_series = [s for s in series_types if constraints_match(s)]
+        assert possible_series, f"No series matches {type_}, {type(frame)}"
+
+        series_strategy = draw(st.sampled_from(possible_series))
+        return draw(series_strategy(type_, frame))
+
+    def value(type_, _frame):
+        return st.builds(Value, value_strategies[type_])
+
+    def select_column(type_, frame):
+        column_names = [n for n, t in schema.column_types if t == type_]
+        return st.builds(SelectColumn, st.just(frame), st.sampled_from(column_names))
 
-    many_rows_per_patient_frame = st.deferred(
-        lambda: st.one_of(
-            select_table,
-            sorted_frame,
-            filter_,
-        )
-    )
+    @st.composite
+    def eq(draw, _type, frame):
+        type_ = draw(any_type())
```

Comment:
The type that's passed to one of these series strategy functions is the type of the resulting series.

In the case of `eq()` that is always `bool`. We could reasonably `assert _type == bool`. But as an internal detail `eq()` needs to pick an arbitrary type for its inputs.

In the case of `add()` the type of the series can vary and is the same as the type of its inputs.

One of the less satisfactory things about this implementation is the information about each series type is smeared out between the strategy functions (which specify the types of the inputs) and the `series_constraits` map in `series()` (which specifies the types of the resulting series).

Reply:
Let me have a crack at answering to check that I've understood this all correctly. I think this is because the `Eq` operation always returns a `bool` (so we don't care about the type we're passed because it's always going to be `bool`), but it accepts arguments of _any_ type (as long as they're the same type).

Whereas `Add` can return either a `float` or an `int` and, if it's a `float`, then its arguments have to be `float`, and if an `int` then its arguments have to be `int`. So it needs to respect the type it was passed.

Does that makes sense?

Reply:
Well, I think Dave's answer is a bit clearer than mine -- so that's a good sign.

Reply:
(I didn't realise you'd already replied when I wrote that, so it was an independent attempt at answering rather than an attempted clarification. But either way, glad to know I've got the fundamentals here.)